File: cpp/209.cpp
#include <vector>
#include <numeric>
using namespace std;

class Solution {
public:
  int minSubArrayLen(int s, vector<int>& nums) {
    if(accumulate(nums.begin(), nums.end(), 0) < s)
      return 0;
    int n = nums.size();
    int mi = n, cur = 0, i = 0, j = 0;
    while(i < n){
      while(j < n && cur < s)
        cur += nums[j++];
      if(cur >= s && j-i < mi)
        mi = j-i;
      cur -= nums[i++];
    }
    return mi;
  }
};

File: python/209.py
class Solution(object):
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        if sum(nums) < s:
            return 0
        n = len(nums)
        mi = n
        i = j = 0
        cur = 0
        while i < n:
            while j < n and cur < s:
                cur += nums[j]
                j += 1
            if cur >= s and j-i < mi:
                print(i, j, cur)
                mi = j-i
            cur -= nums[i]
            i += 1
        return mi

sol = Solution()
print(sol.minSubArrayLen(7, [2,3,1,2,4,3]))
